discovery.md

```diff
@@ -650,30 +630,49 @@ upon return of the `202 Accepted` from the original request.
+Whether an `id` is given or supplied, any existing Service definition with that
+`id` MUST be replaced by the Service definition in the request and any assigned
+values. If an `epoch` is given in the request it MUST be greater than the
+`epoch` of any existing Service of the same `id` or the request MUST fail.
+
+If the Discovery Endpoint is unable to successully add every Service in the
+incoming request then an error MUST be generated and none of the specified
+Services are to be added or updated in the Discovery Endpoint.
+
+The Services in the request MUST be processed in the order in which they are
+given. While any error during the update MUST cause the rollback of any effect
+of the request, the processing of each Service in the request MUST be
+applied as if all previous Services in the request were already completed.
+This means that attribute uniqueness checking, such as on the `name`
+attribute, MUST be done on the new state of all Services as if all previously
+specified Services in the incoming request were already successfully processed.
 
-If any Service specified in the request matches the `name` of any existing
-Service, or the same `name` is used more than once within the request, then
-an error MUST be generated.
+For example, if a Discovery Endpoint has a Service called `MyService`, it is
+possible to rename that Service to `YourService` and then to add, or update,
+another Service to use `MyService` as its `name`. But this is only true if
+the rename of the first Service happens first the request message.
 
-If the Discovery Endpoint is unable to successully add all of the Services
-in the incoming request then an error MUST be generated and none of the
-specified Services are to be added to the Discovery Endpoint.
+Likewise, while it might be less than optimal, it is technically possible for
+a request to update a Service multiple times if the same `id` is used more
+than once within a request.
 
 The follow responses are defined by this specification:
-- `201 Created` if all the specified Services were processed successfully.
-  - If there was only one Service in the incoming request, then the response
-    MUST include an HTTP `Location` Header with a value that points to the
-    correspdonding Service.
-  - If the request contained more than one Service, then the HTTP `Location`
-    Header MUST NOT appear in the response.
-  - The HTTP Response Body MUST include an array of `id` values whose order
-    and values MUST match the list of Services in the request.
+- `200 OK` if all the specified Services were processed successfully
+  - The HTTP Response Body MUST include an array of `id` and `epoch` values
+    whose order and values MUST match the list of Services in the request
+- `400 Bad Request` if the first error encountered is a constraint failure
+- `409 Conflict` if the first error encountered is that a given epoch was less
+  than or equal to the existing epoch of the Service with a given `id`
```

Comment:
here, as elsewhere, making the behavior associated with handling errors specified by this document explicit and standard.